Rails 7 update

mjankowski opened the pull request:
I had this WIP branch going and just rebased/updated it. I was able to get all specs passing in this branch ... but it is by no means ready to go because there are open questions/issues around the changes I made to accomplish that.

The actual bundle update process was straightforward and required bumping all of: rails, rails-i18n, nsa, strong_migrations.

I think rails and rails-i18n are fine, but questions about others:

- `nsa` seems unmaintained, and apparently causes perf issues when used ... should we pull this out? If so, can we pull out of main now so it's not part of the rails upgrade?
- `strong_migrations` needs to go to at least v0.8 to support rails 7, there's a conversation about updating is outside rails 7 here - https://github.com/mastodon/mastodon/pull/24056 - ideally we could bump it prior to rails 7 update as well
- `devise-two-factor` has a pretty complex path to migrate from version 4 to 5 (requires migration, data changes, more migrations, multiple deployments, introduction of rails credentials for secrets, etc). Fortunately there are some PRs open to support both devise-two-factor and attr_encrypted for rails 7 ... so I think we will be able to update to rails 7 first, and then come in later and do the devise-two-factor upgrade.

(This PR originally started to do this devise-2fa migration because it seemed required. That appears to not be the case).

After gem update I ran `rails app:update` and tried to make sensible choices about what changes to keep/reject. Feedback appreciated on that front if I've missed something. Misc small changes:

- The method signature for Rails Preloader changed, easy enough to fix.
- Rails now adds version info (similar to migrations) to the schema as well, seems fine. strong_migrations plays nice with this as well.
- There was an issue with the way we were setting the `response['Link']` headers in some before_action callbacks. I moved these to `after_action` and it resolved the issue. More notes in commit.
- Rails requires an explixit `allow_other_host` now on redirects, added this to a few places the specs were exposing as needing this ... might be more though.

Open issues:

- There are rails changes to an area we monkey patched, and I don't fully understand that area of code. I left some comments on the PR to understand further - https://github.com/mastodon/mastodon/pull/17693#issuecomment-1482843193
- The `select_all` method signature changed. I copy/pasted a random blog post about how to fix this into my branch, and need to understand this further to make sure change is correct.

There are misc TODO comments throughout code here where I know things are not ready yet. I also haven't bothered with rubocop compaints since a lot of this is WIP.

I may open separate PRs for the portions of this which I think could be changed in advance of rails 7 update.

mjankowski committed:
Update rails to version 7.0.6

mjankowski committed:
Run `rails app:update` with version 7

After running, pick through changes and keep sensible defaults while
preserving previous intentional deviations.

mjankowski committed:
Update method signature for AR Preloader

The method now takes `records` and `association` keyword args.

mjankowski committed:
Update Status insert patch for Rails 7

Changes the monkey-patched AR Persistence code to match Rails 7 changes.

mjankowski committed:
Switch load_defaults to rails 7

mjankowski commented:
Follow up note here ... https://github.com/tinfoil/devise-two-factor/issues/226 ... sounds like it might be possible to bump to rails 7 without also bumping to devise-two-factor v5. That would make this all much simpler if so. I'll investigate that and update with results.

mjankowski committed:
Use rails 7 to_fs shorthand for formatted strings

mjankowski commented:
This branch is now passing all specs and rubocop checks, but is not yet ready for review because of the open questions.

I opened a few PRs pulling out some things which will need to change for Rails 7, but could just as easily change ahead of time as well so that we can keep the Rails 7 update PR as small as possible and do other stuff first.

mjankowski commented:
The `attr_encrypted` PR was merged in, but is not yet in a tagged release. I've updated this PR to point to the main repo github instead of the pull request branch, and will update again when they tag a release.

skipkayhil commented on config/application.rb:
```diff
@@ -55,7 +55,7 @@
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
-    config.load_defaults 6.1
+    config.load_defaults 7.0
```
It may be a good idea to cut a release before bumping this value. The `7.0` defaults includes a change to `active_support.cache_format`:
https://github.com/rails/rails/blob/4408b67742cb7ee3cdc3b3d850e1a3be205bd4e6/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_0.rb.tt#L108

The safe way to update the cache format is to deploy everything on Rails 7 with the old cache format, and then only changing the format once everything is on Rails 7.

skipkayhil replied:
:+1: Yep, that makes sense to me

mjankowski commented:
The version of `attr_encrypted` which supports rails 7 has been released - https://github.com/attr-encrypted/attr_encrypted/releases/tag/v4.0.0 - but `devise-two-factor` has a hard version requirement which doesn't allow that gem yet.

We currently need to either a) point at the github repo instead of the released gem for attr_encrypted to get working rails 7 behavior in devise-two-factor (what this PR currently does) or b) when devise-two-factor bumps version to allow newer attr_encrypted, update to point to that and drop the github repo reference in Gemfile. Latter is preferable. I've opened a PR on devise-two-factor to initiate that change or get some other approach worked out.

ClearlyClaire commented:
I'm investigating the `_insert_record` situation, and it seems the main difference is that Rails has started passing uninitialized attributes as `ActiveModel::Attribute::FromDatabase` objects rather than leaving them out. I am not sure where and how this is handled in Rails, I'm investigating.

ClearlyClaire commented:
I believe the required change is the following:
```patch
diff --git a/app/models/status.rb b/app/models/status.rb
index 2757497db..4ef99bb38 100644
--- a/app/models/status.rb
+++ b/app/models/status.rb
@@ -393,7 +393,7 @@ class Status < ApplicationRecord
 
   # Hack to use a "INSERT INTO ... SELECT ..." query instead of "INSERT INTO ... VALUES ..." query
   def self._insert_record(values)
-    if values.is_a?(Hash) && values['reblog_of_id'].present?
+    if values.is_a?(Hash) && values['reblog_of_id']&.value.present?
       primary_key = self.primary_key
       primary_key_value = nil
 
@@ -412,7 +412,7 @@ class Status < ApplicationRecord
       # Since we are using SELECT instead of VALUES, a non-error `nil` return is possible.
       # For our purposes, it's equivalent to a foreign key constraint violation
       result = connection.insert(im, "#{self} Create", primary_key || false, primary_key_value)
-      raise ActiveRecord::InvalidForeignKey, "(reblog_of_id)=(#{values['reblog_of_id']}) is not present in table \"statuses\"" if result.nil?
+      raise ActiveRecord::InvalidForeignKey, "(reblog_of_id)=(#{values['reblog_of_id'].value}) is not present in table \"statuses\"" if result.nil?
 
       result
     else
@@ -435,7 +435,7 @@ class Status < ApplicationRecord
     reblog_bind = nil
     values.each do |name, value|
       attr = arel_table[name]
-      bind = predicate_builder.build_bind_attribute(attr.name, value)
+      bind = predicate_builder.build_bind_attribute(attr.name, value.value)
 
       im.columns << attr
       binds << bind
```

I need to investigate more into what was the exact change that caused this and whether/how we should improve on it, but this change should do the trick.

mjankowski committed:
Add note about rails cache format versions in config

mjankowski commented:
Updated with a Rails-7 version of the Status model insert method monkey patch, based on @ClearlyClaire suggestions. Tried to keep it clear on what original was and what we are replacing.

ClearlyClaire commented on app/models/status.rb:
```diff
@@ -403,29 +403,34 @@ def status_stat
 
-    # Since we are using SELECT instead of VALUES, a non-error `nil` return is possible.
-    # For our purposes, it's equivalent to a foreign key constraint violation
-    result = connection.insert(im, "#{self} Create", primary_key || false, primary_key_value)
-    raise ActiveRecord::InvalidForeignKey, "(reblog_of_id)=(#{values['reblog_of_id']}) is not present in table \"statuses\"" if result.nil?
+    if values.empty?
+      im.insert(connection.empty_insert_statement_value(primary_key))
```
I'd avoid this case to streamline the code, as we know for a fact `values` is not empty. Otherwise this looks good.

mjankowski commented:
Updates:

- Simplified the Status _insert_record patch based on claire feedback
- My [devise-two-factor PR](https://github.com/tinfoil/devise-two-factor/pull/237) was merged into 4.x branch, but there's not a gem release yet - updated this PR to point to their 4.x branch instead of my PR branch. I might open this commit up as a separate PR in advance of Rails 7, because I think it's the case for both attr_encrypted and devise-two-factor that the changes here *allow* but do not *require* Rails 7, so we could merge in ahead of time similar to some of the other stuff.

skipkayhil commented on config/environments/production.rb:
```diff
@@ -73,6 +91,16 @@
+  # require "syslog/logger"
+  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new "app-name")
+
+  if ENV["RAILS_LOG_TO_STDOUT"].present?
```
I noticed this is a change from the previous behavior as well. Before this block was always run but now is behind an ENV var.

mjankowski replied:
Good catch ... I think arguably the env var approach is better, but for the purposes of this PR we just want to perform the upgrade and not change any other configuration, to the extent possible.

I'll update to keep the previous behavior.

mjankowski commented:
Rebased against main and updated Rails to version 7.0.5

Open questions at this point are:

- Status of `nsa` gem
- Discussion above re: the cache change format and how to version/release/document against that
- More local QA with this branch?
- More eyes on config diff here to make sure any changes are intentional

mjankowski committed:
Override rails 7 default for open redirect config

ClearlyClaire commented:
> Rebased against main and updated Rails to version 7.0.5

Thanks again for your work on this!

> Open questions at this point are:
> 
>     * Status of `nsa` gem

I think the plan is to drop it completely in favor of OpenTelemetry, but the PR for that is not completely ready yet.

>     * Discussion above re: the cache change format and how to version/release/document against that

In the past we had a few occasions where we outright broke cached data format and told people to run `tootctl cache clear`, so not having a truly zero-downtime update is not unheard of for Mastodon, but I would like to avoid that if possible.

The most obvious option is to have the first Rails 7 Mastodon release use the old cache format, and a subsequent release use the new cache format. That is easy on us and works for any admin who does not skip releases.

I'd like to continue supporting skipping releases with the least amount of friction possible, but I can't think of a solution that wouldn't make things more complicated and error-prone in the normal without-skipping-updates process.

So I guess the best way forward is to have a Rails 7 release early, and enable the new format *later*, and have subsequent releases have a prominent note in the release notes for a while. The release with the new format does not have to be immediately following the first one switching to Rails 7, the more we wait between, the fewer admins will stumble into this, and the more options they'll have if intermediary versions become uninstallable for some reason (e.g. yanked deps).

>     * More local QA with this branch?

I'll try to do this at some point closer to release.

>     * More eyes on config diff here to make sure any changes are intentional

I'll try to do that somewhat shortly!

mjankowski commented:
Sounds good RE: cache format change and release strategy. Spacing out the change in time from the initial Rails 7 upgrade to merging/releasing the cache format change makes sense.

RE: nsa gem - given we want to keep this for now but that the gem appears to be unmaintained, are we ok merging in as-is in this PR? (ie, pointing to a PR branch on a fork). If not, what's next best option? (make our own fork?, something else?)

renchap commented:
@mjankowski we discussed this PR recently, and @ClearlyClaire will have a closer look and test it. Will try to get this into 4.2.0.

For the nsa gem, we decided to drop it entirely. This should be done in a separate PR, feel free to open it if you have some time!

mjankowski commented:
> For the nsa gem, we decided to drop it entirely. This should be done in a separate PR, feel free to open it if you have some time!

Sounds good -- done: https://github.com/mastodon/mastodon/pull/25265

mjankowski commented on config/environments/development.rb:
```diff
@@ -12,13 +14,22 @@
   # Run rails dev:cache to toggle caching.
   if Rails.root.join('tmp', 'caching-dev.txt').exist?
     config.action_controller.perform_caching = true
+    config.action_controller.enable_fragment_cache_logging = true
```
Adds verbosity to logging https://edgeguides.rubyonrails.org/configuring.html#config-action-controller-enable-fragment-cache-logging

mjankowski commented on config/environments/development.rb:
```diff
@@ -41,12 +52,19 @@
   config.active_support.deprecation = :log
 
+  # Raise exceptions for disallowed deprecations.
+  config.active_support.disallowed_deprecation = :raise
```
https://edgeguides.rubyonrails.org/configuring.html#config-active-support-disallowed-deprecation

mjankowski commented on config/environments/development.rb:
```diff
@@ -41,12 +52,19 @@
+  config.active_support.disallowed_deprecation = :raise
+
+  # Tell Active Support which deprecation messages to disallow.
+  config.active_support.disallowed_deprecation_warnings = []
```
Related to setting above, allows us to specify deprecations to not allow - https://edgeguides.rubyonrails.org/configuring.html#config-active-support-disallowed-deprecation-warnings

mjankowski commented on config/environments/development.rb:
```diff
@@ -41,12 +52,19 @@
   config.active_record.migration_error = :page_load
 
+  # Highlight code that triggered database queries in logs.
+  config.active_record.verbose_query_logs = true
```
What it sounds like! Adds verbosity to DB logs - https://edgeguides.rubyonrails.org/configuring.html#config-active-record-verbose-query-logs

mjankowski commented on config/environments/production.rb:
```diff
@@ -19,27 +21,37 @@
   # or in config/master.key. This key is used to decrypt credentials (and other encrypted files).
   # config.require_master_key = true
 
-  ActiveSupport::Logger.new(STDOUT).tap do |logger|
```
This whole block was kept, but moved down in file.

mjankowski commented on config/environments/production.rb:
```diff
@@ -50,6 +62,12 @@
+  # config.active_job.queue_adapter     = :resque
+  # config.active_job.queue_name_prefix = "mastodon_production"
+
+  config.action_mailer.perform_caching = false
```
Enables fragment caching in mailer templates https://edgeguides.rubyonrails.org/configuring.html#config-action-mailer-perform-caching

mjankowski commented on config/initializers/filter_parameter_logging.rb:
```diff
@@ -1,4 +1,8 @@
 # Be sure to restart your server when you modify this file.
 
-# Configure sensitive parameters which will be filtered from the log file.
-Rails.application.config.filter_parameters += [:password, :private_key, :public_key, :otp_attempt]
```
This feature uses partial regex matching, so the below patterns cover everything here.

mjankowski commented on db/schema.rb:
```diff
@@ -10,7 +10,7 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2023_06_05_085711) do
+ActiveRecord::Schema[6.1].define(version: 2023_06_05_085711) do
```
For the narrow purposes of the `7.0` upgrade, I think this is what we want do. I suspect that what will happen is that the next time someone adds a migration post-7.0 we will have some diff noise from bumping this to 7.0 and elsewhere in file with format changes.

renchap commented:
The current `main` branch displays those deprecation warnings:

```
DEPRECATION WARNING: Merging (("accounts"."domain" IS NOT NULL OR "users"."approved" = $1 AND "users"."confirmed_at" IS NOT NULL)) and (("accounts"."domain" IS NULL OR 1=1)) no longer maintain both conditions, and will be replaced by the latter in Rails 7.0. To migrate to Rails 7.0's behavior, use `relation.merge(other, rewhere: true)`. (called from block in accounts_scope at /Users/renchap/dev/mastodon/mastodon/app/controllers/api/v1/directories_controller.rb:28)
```

I do not see related changes here, should it be updated?

mjankowski commented:
> I do not see related changes here, should it be updated?

I just did a bunch of local experimentation here, and yes, it needs to be updated. The rails 7 rewhere behavior is not harmless, and does modify the meaning of the generated query. Probably makes more sense to address that in advance of rails 7 update in another PR. I'll work out something locally and open that.

mjankowski commented:
Rebased against main and bumped rails version to 7.0.5.1.

Will rebase again when that `rewhere` deprecation PR (or similar) is merged.

mjankowski commented:
Updated rails version to 7.0.6, and combined a few related commits into fewer overall.

Still contingent on the rewhere PR merging first.

Gargron commented:
Sorry, didn't see that it was dependent on another PR. Merged now.

mjankowski commented:
Rebased against main again after the rewhere thing was merged, and marking this out of draft and into ready for review.
